### ISSUE TITLE
Controller can't reconcile service classes in TPR mode

### ISSUE THREAD

**Issue by arschles** [kubernetes-incubator/service-catalog#541]
After adding a valid broker, the controller properly fetches catalog and attempts to write it to the API server as a list of `ServiceClass`es. Doing so fails with the following stack trace:

```
I0314 21:42:22.991993       1 controller.go:154] Processing Broker test-broker
I0314 21:42:22.992014       1 controller.go:163] Creating client for Broker test-broker, URL: http://10.0.0.61:8000
I0314 21:42:22.992020       1 controller.go:167] Adding/Updating Broker test-broker
I0314 21:42:22.992400       1 controller.go:175] Successfully fetched 1 catalog entries for Broker test-broker
I0314 21:42:22.992411       1 controller.go:177] Converting catalog response for Broker test-broker into service-catalog API
I0314 21:42:22.992416       1 controller.go:184] Successfully converted catalog payload from Broker test-broker to service-catalog API
I0314 21:42:22.992419       1 controller.go:187] Reconciling serviceClass user-provided-service (broker test-broker)
I0314 21:42:22.993399       1 request.go:1152] body was not decodable (unable to check for Status): no kind "Status" is registered for version "v1"
E0314 21:42:22.993412       1 controller.go:264] Error creating serviceClass user-provided-service from Broker test-broker: the server rejected our request due to an error in our request (post serviceclasses.servicecatalog.k8s.io)
E0314 21:42:22.993417       1 controller.go:189] Error reconciling serviceClass user-provided-service (broker test-broker): the server rejected our request due to an error in our request (post serviceclasses.servicecatalog.k8s.io)
I0314 21:42:22.993433       1 controller.go:321] Updating ready condition for Broker test-broker to False
I0314 21:42:23.002270       1 controller.go:326] Updated ready condition for Broker test-broker to False
I0314 21:42:23.002377       1 controller.go:154] Processing Broker test-broker
I0314 21:42:23.002381       1 controller.go:163] Creating client for Broker test-broker, URL: http://10.0.0.61:8000
I0314 21:42:23.002385       1 controller.go:167] Adding/Updating Broker test-broker
I0314 21:42:23.002782       1 controller.go:175] Successfully fetched 1 catalog entries for Broker test-broker
I0314 21:42:23.002792       1 controller.go:177] Converting catalog response for Broker test-broker into service-catalog API
I0314 21:42:23.002796       1 controller.go:184] Successfully converted catalog payload from Broker test-broker to service-catalog API
I0314 21:42:23.002837       1 controller.go:187] Reconciling serviceClass user-provided-service (broker test-broker)
I0314 21:42:23.003380       1 request.go:1152] body was not decodable (unable to check for Status): no kind "Status" is registered for version "v1"
E0314 21:42:23.003387       1 controller.go:264] Error creating serviceClass user-provided-service from Broker test-broker: the server rejected our request due to an error in our request (post serviceclasses.servicecatalog.k8s.io)
E0314 21:42:23.003392       1 controller.go:189] Error reconciling serviceClass user-provided-service (broker test-broker): the server rejected our request due to an error in our request (post serviceclasses.servicecatalog.k8s.io)
I0314 21:42:23.003404       1 controller.go:321] Updating ready condition for Broker test-broker to False
I0314 21:42:23.012154       1 controller.go:326] Updated ready condition for Broker test-broker to False
```

Note that I reproduced this by _directly_ creating a third party resource broker, not by creating it via the API server. Using this file (called broker.yaml):

```yaml
apiVersion: servicecatalog.k8s.io/v1alpha1
kind: Broker
metadata:
  name: test-broker
spec:
  url: http://10.0.0.61:8000 # this is the URL to the broker
```

I ran this command:

```console
kubectl create --namespace=steward -f broker.yaml
```

**Comment by arschles** [comment]
Update: this appears to happen with the etcd-based API server also. Both were run against the following broker image: quay.io/kubernetes-service-catalog/user-broker:cc922ca

**Comment by vaikas-google** [comment]
Are there any logs on the server side? As in why was it rejected:
the server rejected our request due to an error in our request (post serviceclasses.servicecatalog.k8s.io)

**Comment by arschles** [comment]
@vaikas-google here is the extent of the logs related to `POST`ing on `serviceclasses`:

```console
I0314 22:56:51.033335       1 panics.go:76] POST /apis/servicecatalog.k8s.io/v1alpha1/serviceclasses: (964.942µs) 422 [[controller-manager/v0.0.0 (linux/amd64) kubernetes/$Format/service-catalog-controller-manager] 172.17.0.5:48848]
I0314 22:56:51.035530       1 round_trippers.go:393] curl -k -v -XGET  -H "User-Agent: apiserver/v0.0.0 (linux/amd64) kubernetes/$Format" -H "Authorization: Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJzdGV3YXJkIiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZWNyZXQubmFtZSI6ImRlZmF1bHQtdG9rZW4tZmh2Z2siLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC5uYW1lIjoiZGVmYXVsdCIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50LnVpZCI6IjFiMzA1NWY3LTA4ZjgtMTFlNy1hY2U5LTA4MDAyNzIzMjYyYSIsInN1YiI6InN5c3RlbTpzZXJ2aWNlYWNjb3VudDpzdGV3YXJkOmRlZmF1bHQifQ.QL3SiX6YZPRrFY8iMmcSX4cMO13DtRzYGjBwnOOZ48wpLQKDeH_RBo-nSWbPdAYYp4fhqjUqdx-uIT3QRevwFi2skTfm_vdRNb9kvvfCUvceOedMaR273XPNgawmtYnvMUQ_Da2AB6LeSMlf0BM11bMqAaMqYSpIjBjHfOaaFuZ0LIrBsyF5zNfREJj-hq9SiSRv21tH7TdzWMqrNIWehxQNqzDb6D93jqqYZlEm-g68JAEyEJv0s0CY6YJPsAG8tHofT8ML9JwESeG7l1g5UDjZ2jZ6T669n-8qw1rmvoxUuySamkHL9nMTyaIu3ZtewS4_EccdnotRAazyY00W3Q" -H "Accept: application/json, */*" https://10.0.0.1:443/apis/servicecatalog.k8s.io/v1alpha1/namespaces/steward/brokers/test-broker
```

repro steps incoming

**Comment by arschles** [comment]
Repro steps:

1. Install the UPS broker with the following:

```console
helm install \
	--namespace=steward \
	--set upsBrokerVersion=cc922ca,registry=quay.io/kubernetes-service-catalog,debug=true \
	deploy/ups-broker
```
2. Get the cluster IP of the `ups-broker` service. You'll use it in the next step.
3. Save a `broker.yaml` file with the below contents. Replace BROKER_IP with the IP that you got from the previous step.

```yaml
apiVersion: servicecatalog.k8s.io/v1alpha1
kind: Broker
metadata:
  name: test-broker
spec:
  url: http://BROKER_IP:8000
```
4. Start up the service catalog API and controller with the following command:

```console
helm install \
    --namespace=steward \
    --set version=${VERSION},storageType=tpr,debug=true,insecure=true,imagePullPolicy=Never,globalNamespace=steward \
    deploy/catalog
```
5. Create the broker by running the following command against the service catalog API server: `kubectl create -f broker.yaml`
    - You can also elect to run it against the core API server since step 4 ran the API server in TPR mode. To do that, run `kubectl create -f broker.yaml --namespace=steward`
6. Wait for the controller to pick up the broker create event
7. Get the broker's status by running the following command against the service catalog API server: `kubectl get -o yaml broker test-broker`
    - You can also elect to run it against the core API server since step 4 ran the API server in TPR mode. To do that, run `kubectl get broker -o yaml test-broker --namespace=steward`

After step 7, notice that the broker has an error condition that looks similar to the following:

```yaml
  conditions:
  - message: Error syncing catalog from Broker
    reason: ErrorSyncingCatalog
    status: "False"
    type: Ready
```

Respective logs in the API server and controller seem to indicate a failure to update the service classes that correspond to the broker's catalog.

**Comment by vaikas-google** [comment]
Can you grab the output of the broker by doing a curl from one of the pods in that cluster?
http://10.0.0.61:8000/v2/catalog

It's in controller.go this line (depending on which version you're running) reconcileServiceClassFromBrokerCatalog

c.serviceCatalogClient.ServiceClasses().Create(serviceClass)

**Comment by arschles** [comment]
Possibly related to https://github.com/kubernetes-incubator/service-catalog/issues/412

**Comment by arschles** [comment]
@vaikas-google the broker outputs no logs:

```console
ENG000656:service-catalog aaronschlesinger$ ks logs -f ups-broker-4136751632-7m512

```

**Comment by arschles** [comment]
Also, here is the `/v2/catalog` response:

```console
ENG000656:service-catalog aaronschlesinger$ curl http://192.168.99.100:31420/v2/catalog
{"services":[{"name":"user-provided-service","id":"4F6E6CF6-FFDD-425F-A2C7-3C9258AD2468","description":"User Provided Service","bindable":false,"plan_updateable":false,"tags":null,"requires":null,"metadata":null,"plans":[{"name":"default","id":"86064792-7ea2-467b-af93-ac9694d96d52","description":"User Provided Service","metadata":null,"free":true,"schemas":null}],"dashboard_client":null}]}
```

**Comment by arschles** [comment]
here's another update: I tried running `kubect create -f` with this service class template: 

```yaml
apiVersion: servicecatalog.k8s.io/v1alpha1
kind: ServiceClass
metadata:
  name: test-serviceclass
brokerName: test-broker
bindable: true
osbGuid: D35b55b2-b1fd-4123-8045-5b9c619cb629
plans:
  - name: example-plan-1
    osbGuid: 10E03cb7-b2cf-40dd-a954-16a382b92446
    osbFree: true
  - name: example-plan-2
    osbGuid: b8269ab4-7d2d-456d-8c8b-5aab63b321d1
    osbFree: false
```

Notice that the `osbGUID` has an upper case letter in it, which violates the `NameIsDNS1035Label` validation step (see https://github.com/kubernetes-incubator/service-catalog/pull/547#issue-214256219 for more details on this validation step). The response to this `kubectl create -f` command is the following:

```console
ENG000656:service-catalog aaronschlesinger$ ./kubectl create -f resources/service-class-invalid.yaml 
The ServiceClass "test-serviceclass" is invalid: 
* osbGuid: Invalid value: "D35b55b2-b1fd-4123-8045-5b9c619cb629": must match the regex [a-z0-9]([-a-z0-9]*[a-z0-9])? (e.g. 'my-name' or '123-abc')
* plans[0].osbGuid: Invalid value: "10E03cb7-b2cf-40dd-a954-16a382b92446": must match the regex [a-z0-9]([-a-z0-9]*[a-z0-9])? (e.g. 'my-name' or '123-abc')
```

This means that the server is returning a valid error response, and the client is ignoring the error string somewhere

**Comment by pmorie** [comment]
...and loosened the validations in #558

**Comment by arschles** [comment]
This was fixed by #558 (and debugging was added to aid in the solution & future debuggability in #556). Closing
